send/slack

```diff
@@ -0,0 +1,518 @@
+
+	response = client.patch_user(id=their_user["id"], partial_user=user_data)
+
+	if response.errors is not None and response.errors.description.startswith(
+		"bad_email_address"
+	):
+		WARNING.append(f"Found duplicated emails - user not updated {our_user}")
+		return None
+
+	check_response(response)
+	global USERS_UPDATED
+	USERS_UPDATED += 1
+
+	return response.user
+
+
+def deactivate_user(client, their_user):
+	print_debug(f"Deactivating user {their_user['real_name']}")
+	patch_result = client.patch_user(
+		id=their_user["id"], partial_user=User(active=False)
+	)
+	if patch_result.errors is not None and (
+		"cannot_modify_owner" in patch_result.errors.description
+	):
+		WARNING.append(f"Cannot deactivate {their_user.user_name}")
+		return
+
+	check_response(patch_result)
+	global USERS_DEACTIVATED
+	USERS_DEACTIVATED += 1
+
+
+def get_their_users(client):
+	users = []
+	while True:
+		response = client.search_users(
+			start_index=len(users) + 1,
+			count=COUNT,
+		)
+		check_response(response)
+		users.extend(response.users)
+
+		if len(response.users) < COUNT:
+			break
+	return users
+
+
+def update_users(client, our_data, their_users, skip_users):
+	"""
+	Users are matched by username
+	"""
+	our_users = our_data.get("users")
+	our_usernames = our_users.keys()
+	deactivated_users = []
+
+	for their_user in their_users:
+		if their_user["name"] in skip_users:
+			continue
+		if (
+			their_user["is_bot"]
+			or their_user["deleted"]
+			or their_user["is_restricted"]
+			or their_user["is_ultra_restricted"]
+			or their_user["is_app_user"]
+		):
+			continue
+
+		our_user = next((u for u in our_usernames if u == their_user["name"]), None)
+		# deactivate users (except bots, guests ...)
+		if our_user is None and not their_user["deleted"]:
+			deactivate_user(client, their_user)
+			deactivated_users.append(their_user["id"])
+
+	for our_username in our_usernames:
+		if our_username in skip_users:
+			continue
+		our_user = our_users[our_username]
+		our_user["userName"] = our_username
+		their_user = next((u for u in their_users if u["name"] == our_username), None)
+
+		# create missing users
+		if their_user is None:
+			created_user = create_user(client, our_user)
+			if created_user is not None:
+				users_mapping[created_user.user_name] = created_user.id
+				continue
+
+		# update users
+		elif (
+			our_user["firstName"] != their_user["profile"]["first_name"]
+			or our_user["lastName"] != their_user["profile"]["last_name"]
+			or our_user["email"] != their_user["profile"]["email"]
+			or their_user["deleted"]
+			or their_user["is_restricted"]
+			or their_user["is_ultra_restricted"]
+		):
+			update_user(client, our_user, their_user)
+
+	return deactivated_users
+
+
+def create_channel(webclient, name, is_private):
+	print_debug(f"Creating {'public' if not is_private else 'private'} channel {name}")
+	try:
+		response = webclient.conversations_create(name=name, is_private=is_private)
+	except SlackApiError as e:
```

Comment:
I don't think this ever gets raised (according to docs https://slack.dev/python-slack-sdk/api-docs/slack_sdk/#slack_sdk.WebClient.conversations_create). Also, why check for exception here and the `error` property in `invite_members`?

Reply:
i think the two approaches are the same, but i unified them.